### ISSUE TITLE
How to get english wikipedia images (thumbnails that appear on article only)?

### ISSUE THREAD

**Issue by NijazM** [gnosygnu/xowa#764]
I read all the help but I am still not sure. Should I just download commons.wikimedia.org? If yes, which file exactly? Put it where then?
Should I use command line? I don't need full size images, but only those that appear on articles, before clicking on them. Should I have strong hardware? Do I really need 22 TB? Why then commons dump files take much less space and even kiwix whole wikipedia with images (but without few namespaces like category and wikipedia are not included) takes only 89 GB. That is much less than 1 TB. When I click on some image it says:
```

This offline page is a reduced version of the online one: https://commons.wikimedia.org/wiki/File:Potato_plant_prior_to_harevest.jpg.

If you want XOWA to show an offline page just like the online version, you should download commons.wikimedia.org at Import online.

    This wiki will use at least another 20 GB of disk space.
    This wiki only provides wikitext. No extra images are downloaded. Note that the total size of images for commons.wikimedia.org would be too large (approximately 22 TB).

```

**Comment by gnosygnu** [comment]
>  Should I just download commons.wikimedia.org? 

You only need `commons. wikimedia.org` if you want to visit pages in `https://commons.wikimedia.org`. Most users don't actually go there; they go to pages like `https://en.wikipedia.org`.

> If yes, which file exactly? Put it where then? Should I use command line? 

I'd start off by downloading through home/wiki/Dashboard/Import/Online. When done, it will generate in `C:\xowa\wiki\commons.wikimedia.org` (or wherever you installed XOWA)

> I don't need full size images, but only those that appear on articles, before clicking on them.

For enwiki thumbnails (images appearing in articles), you really should wait for me to upload the 2020-06 version. I'm almost done and they should be uploaded within a week or so.

If you want to get the 2020-02 version, you can download them from here: https://archive.org/details/Xowa_enwiki_2020-02

> Should I have strong hardware?

Only if you plan to build your own version of enwiki. This is pretty complicated though, so I'd recommend waiting for my upload.

> Do I really need 22 TB?

22 TB is for all the full size images in commons.wikimedia.org. This 22 TB is just an estimate; I've never done this.

> Why then commons dump files take much less space and even kiwix whole wikipedia with images (but without few namespaces like category and wikipedia are not included) takes only 89 GB. 

For enwiki, XOWA clocks in at around 160 GB. The extra space is partly due Portal / Category / Wikipedia namespaces and partly due to XOWA's database implementation. (this is a bit complicated to explain)

> That is much less than 1 TB. When I click on some image it says:

Yeah, again this excerpt is for downloading full images. For now, you'd only want thumbnail ones.

**Comment by AiliAili** [comment]
Hi gnosygnu,
Could I directly download from https://archive.org/details/Xowa_enwiki_2020-02
and directly extract HTML files from the downloaded files without preprocessing?
Thanks.

**Comment by NijazM** [comment]
Thanks for explaining clearly. Just one more question, should I download all files or are there some duplicates or not needed ones? What are those with "html" in file name?

**Comment by gnosygnu** [comment]
> Could I directly download from https://archive.org/details/Xowa_enwiki_2020-02
> and directly extract HTML files from the downloaded files without preprocessing?

Yup. All the HTML is in the `.html.xowa` databases. Keep in mind they are gzip'd, so you would need to have some way of running `gunzip` or other common gzip decompressor

**Comment by gnosygnu** [comment]
> Thanks for explaining clearly.

No problem!

> Just one more question, should I download all files or are there some duplicates or not needed ones? 

There are no duplicates. You could skip all the `.text.xowa` databases: they only contain wikitext and would be needed only if you are editing your offline copy

> What are those with "html" in file name? 

These are the main databases with the HTML content.

> Also can I only download some files and combine them with my already existing files? Will it work? 

It will work, but the results will be a little off. Basically, your current version (assuming it's 2020-07) would be looking for new files that could not exist in the 2020-02 version

> I already have all except thumbnails. Then I just have to download "core" and "file" files?

Almost. You'd only download all the files starting with `Xowa_enwiki_2020-02_file_`.
* Start with `Xowa_enwiki_2020-02_file_core.zip`
* End with `Xowa_enwiki_2020-02_file_ns.100_db.001.zip`

Hope this helps!

**Comment by AiliAili** [comment]
Hi gnosygnu,

I downloaded the html databases, suppose it is "en.wikipedia.org-html-ns.000-db.022.xowa"
I queried the HTML for page_id=60566296, using "sqlite3 en.wikipedia.org-html-ns.000-db.022.xowa  "SELECT body from html where page_id=60566296;"  > tem.html"
But only returned ‹, not a webpage as I expected.
Am I doing sth wrong here?

**Comment by gnosygnu** [comment]
> I queried the HTML for page_id=60566296, using "sqlite3 en.wikipedia.org-html-ns.000-db.022.xowa "SELECT body from html where page_id=60566296;" > tem.html"

Thanks for the command-line snippet. :+1: 

> Am I doing sth wrong here?

You're missing a step. Quoting from above:

>  Keep in mind they are gzip'd, so you would need to have some way of running gunzip or other common gzip decompressor

If you're on Windows, you can run something like the below:

```
D:\xowa\bin\windows\sqlite\sqlite3.exe D:\xowa\wiki\en.wikipedia.org\en.wikipedia.org-html-ns.000.xowa "SELECT hex(body) from html where page_id=303;" | xxd -r -p > D:\xowa\temp.gz & D:\xowa\bin\windows\7-zip\7za.exe e D:\xowa\temp.gz -y & mv D:\xowa\temp D:\xowa\temp.html
```

Hope this helps.

**Comment by AiliAili** [comment]
Thank you for your timely reply.
Actually, I am using ubuntu.
Do you mean the retrieved html from the table is compressed? So I need to decrepress it first to get the rendered HTML I actual want.

**Comment by gnosygnu** [comment]
> Do you mean the retrieved html from the table is compressed? So I need to decrepress it first to get the rendered HTML I actual want.

Yup. The HTML in the table is compressed with gz. Just pipe sqlite to gunzip to get the actual HTML

**Comment by NijazM** [comment]
It is still not completely clear to me what are files with html in file name? You say they are _These are the main databases with the HTML content._
But how come I have not them? Are they needed for images too?
I only have these files, what is the difference between having and not having html?
```
en.wikipedia.org-core.xowa
en.wikipedia.org-file-core.xowa
en.wikipedia.org-file-user.xowa
en.wikipedia.org-text-ns.000.xowa
en.wikipedia.org-text-ns.000-db.002.xowa
en.wikipedia.org-text-ns.000-db.003.xowa
en.wikipedia.org-text-ns.000-db.004.xowa
en.wikipedia.org-text-ns.000-db.005.xowa
en.wikipedia.org-text-ns.000-db.006.xowa
en.wikipedia.org-text-ns.000-db.007.xowa
en.wikipedia.org-text-ns.000-db.008.xowa
en.wikipedia.org-text-ns.000-db.009.xowa
en.wikipedia.org-text-ns.000-db.010.xowa
en.wikipedia.org-text-ns.000-db.011.xowa
en.wikipedia.org-text-ns.000-db.012.xowa
en.wikipedia.org-text-ns.004.xowa
en.wikipedia.org-text-ns.004-db.002.xowa
en.wikipedia.org-text-ns.004-db.003.xowa
en.wikipedia.org-text-ns.006.xowa
en.wikipedia.org-text-ns.008.xowa
en.wikipedia.org-text-ns.010.xowa
en.wikipedia.org-text-ns.012.xowa
en.wikipedia.org-text-ns.014.xowa
en.wikipedia.org-text-ns.100.xowa
en.wikipedia.org-text-ns.108.xowa
en.wikipedia.org-text-ns.118.xowa
en.wikipedia.org-text-ns.710.xowa
en.wikipedia.org-text-ns.828.xowa
en.wikipedia.org-text-ns.2300.xowa
en.wikipedia.org-xtn.category.core.xowa
en.wikipedia.org-xtn.category.link-db.001.xowa
en.wikipedia.org-xtn.category.link-db.002.xowa
en.wikipedia.org-xtn.category.link-db.003.xowa
en.wikipedia.org-xtn.category.link-db.004.xowa
en.wikipedia.org-xtn.category.link-db.005.xowa
en.wikipedia.org-xtn.category.link-db.006.xowa
en.wikipedia.org-xtn.category.link-db.007.xowa
en.wikipedia.org-xtn.category.link-db.008.xowa
en.wikipedia.org-xtn.category.link-db.009.xowa
en.wikipedia.org-xtn.category.link-db.010.xowa
en.wikipedia.org-xtn.category.link-db.011.xowa
en.wikipedia.org-xtn.category.link-db.012.xowa
en.wikipedia.org-xtn.category.link-db.013.xowa
en.wikipedia.org-xtn.search.core.xowa
en.wikipedia.org-xtn.search.link-title-ns.000-db.001.xowa
en.wikipedia.org-xtn.search.link-title-ns.999-db.001.xowa
```

**Comment by gnosygnu** [comment]
> It is still not completely clear to me what are files with html in file name? You say: These are the main databases with the HTML content.

There are 2 types of databases for pages:
* `*-text-*.xowa`: These databases have the wikitext -- the "source code" for a page. For an example of wikitext, see this version for Earth: https://en.wikipedia.org/w/index.php?title=Earth&action=edit
* `*-html-*.xowa`:These databases have the HTML -- the actual page itself. For an example of HTML, see this version for Earth: `view-source:https://en.wikipedia.org/wiki/Earth` (put this url into Chrome).

So, XOWA works like this:
* It downloads a huge XML file from the Wikimedia dumps (For example: https://dumps.wikimedia.org/enwiki/20200720/). This has the wikitext ("source code") for every page in a wiki
* It converts this huge XML file into separate `*-text-*.xowa` files. It does this b/c it's quicker to work with a database than a huge XML file.
* Whenever you visit a page in XOWA, XOWA retrieves the wikitext for the page and builds HTML for it. This HTML is what is shown in the XOWA GUI (or the browser if you're using HTTP Server)

Now sometimes it takes some time to generate HTML from wikitext. Using an example like the Earth page:
* On a fast desktop, it can take up to 10 seconds
* On a slower desktop, it can take 30 seconds
* On a Raspberry Pi, it can take a minute or so
* On Android, it will take upwards of 10 minutes.

So, to skip this building phase, I also provide `*-html-*.xowa` databases. These have precompiled versions of HTML, so they will load very quickly (less than 1 second). These `*-html-*.xowa` databases take a long time to build since it involves running XOWA against every page in a wiki.

> Also what is included exactly in you file files? Images only or video and audio too? What about gif?

Only images are included (no audio / video).

Also, generally only "thumbnail" images. A thumbnail is the image as shown in the page, not the "original" image (which is much larger). For an example:
* Thumbnail: https://upload.wikimedia.org/wikipedia/commons/thumb/9/97/The_Earth_seen_from_Apollo_17.jpg/599px-The_Earth_seen_from_Apollo_17.jpg
* Original: https://upload.wikimedia.org/wikipedia/commons/9/97/The_Earth_seen_from_Apollo_17.jpg

> Also how can I make word search database? I imported search but looks like that is only title search? See files below, some are missing?

The search database is "full-text search". You can build it using the instructions here: http://xowa.org/home/wiki/App/Full-text_search/Lucene/Search_indexes/Building

> But how come I have not them? Are they needed for images too?

Assuming you're asking about text vs html again, no, not needed for images

> I only have these files, what is the difference between having and not having html?

So, here are the scenarios:
* `-text-.xowa` only: Can edit pages. Can view pages, but may need to wait several seconds to view
* `-html-.xowa` only: Can not edit pages. Can view pages, and pages will load sub-second
* `-text-.xowa` and `-html-.xowa`: Can edit pages and view pages in sub-second time. Will take more space.

Hope this helps

**Comment by NijazM** [comment]
Now I understand. Just two more questions.
Do html take more space than text?
What file types are included in images? Are animated gifs and svg?

**Comment by gnosygnu** [comment]
> Do html take more space than text?

Yup
* text db: 24.4 GB
* html:  45.3 GB

> What file types are included in images? Are animated gifs and svg?

All images. SVGs though are converted to PNG.

**Comment by NijazM** [comment]
Is there quick and easy way to download all images of only few articles, without having to make or download all images?
For example of only 10 articles or whole category or template. At least like manually watching page while online via disk cache.
But without having to open each page manually. I already easily mark and copy all links of some template or category via server and using firefox extension 'copy selected links'. Also I can easily search and replace beginning or any part of all such links via notepad++ if needed.

**Comment by gnosygnu** [comment]
> Is there quick and easy way to download all images of only few articles, without having to make or download all images?

Nope, unfortunately not. The closest thing you can do is load several articles at once using a keyboard shortcut. Specifically:
* Open up XOWA GUI
* Go to en.wikipedia.org/wiki/Main_Page
* Copy the following snippet. Note that it is 3 article titles with each title on its own line:
```
Earth
Sun
Mars
```
* Click the URL bar
* Press `Ctrl+Enter`. This will take the last clipboard entry and open a tab for each line. (In this case, you'll get 3 new tabs)

Basically, this automates opening up a page one at a time. Keep in mind that each article will open on a separate thread, so if you have 50 or so articles, it could take several minutes for all 50 to complete.

> Also, can I convert wikitext database to html? How? And can I remove wikitext then?

Nope, not easily. The only way to do it is to manually build the html databases (which is what I do). The process is fairly complicated and takes a long time (10 days for English Wikipedia). If you're intersted, see http://xowa.org/home/wiki/Dev/Command-line/Dumps
